File: 05/java/Quadratic.java
/* Quadratic.java computes the roots of a quadratic equation
 * Completed by: Jonathan Winkle
 * 3/10/20
 ***************************************************************/

import java.util.Scanner;
import java.lang.Math;

// java class Quadratic
public class Quadratic
{
  // main program
  public static void main(String[] args)
    {
      Scanner keyboard = new Scanner(System.in);
      // prompt for the string
      System.out.println("\nTo compute the roots of a quadratic formula in the form y = ax^2 + bx + c, enter a, b, and c: ");
      double a = keyboard.nextDouble();
      double b = keyboard.nextDouble();
      double c = keyboard.nextDouble();
      
      //Create list, compute results and print
      double [] resultList = new double[2];
      
      if (quadraticRoots(a, b, c, resultList)) {
        System.out.println("\nThe first root is " + resultList[0] + "\n and the second root is " + resultList[1]);
      } else {
        System.out.println("\nUnable to find roots");
      }
    }
  
  /* quadraticRoots() computes the roots of a quadratic equation
   * Receive: 	a, b, c,
   *            results - an array of doubles in which to place roots
   * Return: true if roots are valid, false if not valid
   *****************************************************/
  public static boolean quadraticRoots(double a, double b, double c, double[] results)
  {
    if (a != 0) {
        double arg = Math.pow(b, 2.0) - (4 * a * c);
        if (arg >= 0) {
            results[0] = (-b + Math.sqrt(arg))/(2*a);
            results[1] = (-b - Math.sqrt(arg))/(2*a);
            return true;
        } else {
            System.out.println("\nb^2 - 4ac is negative!");
            return false;
        }
    } else {
        System.out.println("\na is zero!");
        return false;
    }
  } 


}

File: 05/ruby/quadratic.rb
# quadratic.rb computes the roots of a quadratic equation.
#
# Completed by: Jonathan Winkle
# Date: 3/10/20
#################################################################


# quadform() computes the roots of a quadratic equation.
# Receive: a, b, c, the coefficients, vec                 
# PRE: a >= 1, b^2 - 4ac >= 0
# Return: true if the roots are valid, false if not
######################################################

def quadform (a, b, c, vec)
    if a != 0 then 
        arg = b**2 - (4*a*c)
        if arg >= 0 then
            vec[0] = (-b + Math.sqrt(arg))/(2*a)
            vec[1] = (-b - Math.sqrt(arg))/(2*a)
            true
        else
            puts "b^2 - 4ac is negative!"
            false
        end
    else
        puts "a is zero!"
        false
    end
  end
  
  
  def main
    print "To compute the roots of a quadratic formula in the form y = ax^2 + bx + c, enter a, b, and c: "; 
    a = gets.to_f 
    b = gets.to_f
    c = gets.to_f
    results = Array.new(2)
    if quadform(a, b, c, results) then
        puts "The first root is: " << results[0].to_s
        puts "the second root is: " << results[1].to_s
    else
        puts "Unable to compute the roots"
    end
  end

  main